hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/index/ScheduleIndexActionExecutor.java

```diff
@@ -0,0 +1,138 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.apache.hudi.common.model.WriteConcurrencyMode.OPTIMISTIC_CONCURRENCY_CONTROL;
+import static org.apache.hudi.config.HoodieWriteConfig.WRITE_CONCURRENCY_MODE;
+
+/**
+ * Schedules INDEX action.
+ * <li>
+ * 1. Fetch last completed instant on data timeline.
+ * 2. Write the index plan to the <instant>.index.requested.
+ * 3. Initialize file groups for the enabled partition types within a transaction.
+ * </li>
+ */
+public class ScheduleIndexActionExecutor<T extends HoodieRecordPayload, I, K, O> extends BaseActionExecutor<T, I, K, O, Option<HoodieIndexPlan>> {
+
+  private static final Logger LOG = LogManager.getLogger(ScheduleIndexActionExecutor.class);
+  private static final Integer INDEX_PLAN_VERSION_1 = 1;
+  private static final Integer LATEST_INDEX_PLAN_VERSION = INDEX_PLAN_VERSION_1;
+
+  private final List<MetadataPartitionType> partitionsToIndex;
+  private final TransactionManager txnManager;
+
+  public ScheduleIndexActionExecutor(HoodieEngineContext context,
+                                     HoodieWriteConfig config,
+                                     HoodieTable<T, I, K, O> table,
+                                     String instantTime,
+                                     List<MetadataPartitionType> partitionsToIndex) {
+    super(context, config, table, instantTime);
+    this.partitionsToIndex = partitionsToIndex;
+    this.txnManager = new TransactionManager(config, table.getMetaClient().getFs());
+  }
+
+  @Override
+  public Option<HoodieIndexPlan> execute() {
+    // validate partitionsToIndex
+    if (!EnumSet.allOf(MetadataPartitionType.class).containsAll(partitionsToIndex)) {
+      throw new HoodieIndexException("Not all partitions are valid: " + partitionsToIndex);
+    }
+    // ensure lock provider configured
+    if (!config.getWriteConcurrencyMode().supportsOptimisticConcurrencyControl() || StringUtils.isNullOrEmpty(config.getLockProviderClass())) {
+      throw new HoodieIndexException(String.format("Need to set %s as %s and configure lock provider class",
+          WRITE_CONCURRENCY_MODE.key(), OPTIMISTIC_CONCURRENCY_CONTROL.name()));
+    }
+    // make sure that it is idempotent, check with previously pending index operations.
+    Set<String> indexesInflightOrCompleted = Stream.of(table.getMetaClient().getTableConfig().getInflightMetadataIndexes().split(","))
+        .map(String::trim).filter(s -> !s.isEmpty()).collect(Collectors.toSet());
+    indexesInflightOrCompleted.addAll(Stream.of(table.getMetaClient().getTableConfig().getCompletedMetadataIndexes().split(","))
+        .map(String::trim).filter(s -> !s.isEmpty()).collect(Collectors.toSet()));
+    Set<String> requestedPartitions = partitionsToIndex.stream().map(MetadataPartitionType::getPartitionPath).collect(Collectors.toSet());
+    requestedPartitions.removeAll(indexesInflightOrCompleted);
+    if (!requestedPartitions.isEmpty()) {
```

Comment:
So this assumes, we drop the existing partitions if we need to recreate

Reply:
Originally in my mind, I only had one table config, that tracks "completed" partitions/indexes (in this PRs terminology) and the timeline is the source of truth for whats inflight. how do we handle the scenario where we fail after updating  the tableConfig/hoodie.props, but before writing the requested indexing to the timeline.

Reply:
> how do we handle the scenario where we fail after updating the tableConfig/hoodie.props, but before writing the requested indexing to the timeline.

Case 1: scheduling index
No table config is updated while scehduling. What could happen here is for e.g. column_stats partition was initialized and executor failed just before writing the requested indexing to the timeline. When scheduling is re-triggered, it will begin with a new instant and redo the whole thing. So, this isn't truly idempotent. I need to add a check to ignore if initialization was complete.

Case 2: building index
"completed" partitions table config gets updated only after index catchup and partition is fully built out. If the table config gets updated but executor failed just before writing the completed indexing to the timeline, the regular writers will see that the partition is availbale for updates. There will be an inflight indexing instant in the timeline forever. When indexer is re-triggered, it will fail again because it will that there is already an inflight indexing instant with the same timestamp.

Reply:
Don't we need to handle case 2?

Reply:
Meta point was - why do we need to track the inflight partitions in table metadata?